### PR TITLE
X-Ray traceid extraction - 64bit support

### LINKED ISSUE #3075: Enhance X-Ray propagator to support 64 bits traces extraction
**Is your feature request related to a problem? Please describe.**
Current X-Ray propagator (https://github.com/open-telemetry/opentelemetry-java/blob/main/extensions/aws/src/main/java/io/opentelemetry/extension/aws/AwsXrayPropagator.java) is very strict, requiring full 128 bit trace Id for extraction. In particular, second part of the ID has to be 8 chars long. The problem arises with 64 bit traces, padded with 0s, that get compressed by AWS. 

Consider following scenario:
- B3 configured as propagator for a service
- 64 bit traced id coming with request (`795bf8769fe8b9c0`), extracted, padded with 0s - OTEL internal 128 bit representation
- trace Id injected into outgoing AWS SDK request (say - SQS), second part (8 chars) - all 0s - `Root=1-00000000-00000000795bf8769fe8b9c0;Parent=7a339e56d8f86bb1;Sampled=1`
- trace id extracted upon consuming SQS message - first part (0s) compressed by AWS to a single 0 - `Root=1-0-00000000795bf8769fe8b9c0;Parent=7a339e56d8f86bb1;Sampled=1`
- AwsXrayPropagator unable to recognize the trace id upon injection, new trace id is created - propagation broken

**Describe the solution you'd like**
Enhance AwsXrayPropagator to recognize "compression" of the second part as done by AWS. In such case the ID should stil be recognized as valid and left-padded with 0s.

**Describe alternatives you've considered**
Filing issue towards AWS to disable "compression" of leading 0s.

**Additional context**
Tested with current agent and AWS infra, leading 0s are always compressed (removed). 

Tagging @anuraaga for AWS visibility :) If the proposal makes sense, should be relatively easy to implement.

### PR DESCRIPTION
**Pull request by kubawach**
Closes #3075
Now epoch part of the X-Ray can be between 1 and 8 chars, which allows 64bits with truncated leading 0s to be extracted successfully.

### CODE REVIEW

File: extensions/aws/src/main/java/io/opentelemetry/extension/aws/AwsXrayPropagator.java [hunk 1]
```diff
@@ -210,16 +211,20 @@ private static String parseTraceId(String xrayTraceId) {
     }
 
     // Check delimiters
-    if (xrayTraceId.charAt(TRACE_ID_DELIMITER_INDEX_1) != TRACE_ID_DELIMITER
-        || xrayTraceId.charAt(TRACE_ID_DELIMITER_INDEX_2) != TRACE_ID_DELIMITER) {
+    int firstDelimiter = xrayTraceId.indexOf(TRACE_ID_DELIMITER);
```
Comment (anuraaga) [inline extensions/aws/src/main/java/io/opentelemetry/extension/aws/AwsXrayPropagator.java:214]: Can we keep the original code and add a parseSlow for the short case? It's against spec so not fair all parsing gets a bit slower :)

Comment (kubawach) [reply extensions/aws/src/main/java/io/opentelemetry/extension/aws/AwsXrayPropagator.java:239]: Definitely! Please have a look at the updated version.

File: extensions/aws/src/test/java/io/opentelemetry/extension/aws/AwsXrayPropagatorTest.java [hunk 2]
```diff
@@ -276,6 +276,85 @@ void extract_nullGetter() {
...
+        .isEqualTo(
+            SpanContext.createFromRemoteParent(
+                "00000000d188f8fa79d48a391a778fa6",
+                SPAN_ID,
+                TraceFlags.getSampled(),
+                TraceState.getDefault()));
+  }
+
+  @Test
+  void extract_epochParTwoChars() {
```
Comment (anuraaga) [inline extensions/aws/src/test/java/io/opentelemetry/extension/aws/AwsXrayPropagatorTest.java:296]: Think you meant part

**Review by anuraaga — APPROVED** [review]
Thanks!

File: extensions/aws/src/test/java/io/opentelemetry/extension/aws/AwsXrayPropagatorTest.java (hunk 2 above)
Comment (kubawach) [reply extensions/aws/src/test/java/io/opentelemetry/extension/aws/AwsXrayPropagatorTest.java:296]: Indeed :D

**Review by jkwatson — APPROVED** [review]
Would just like an scare-comment in the StringUtils about it being internal, then good to merge.

File: api/all/src/main/java/io/opentelemetry/api/internal/StringUtils.java [hunk 3]
```diff
@@ -5,12 +5,73 @@
 
 package io.opentelemetry.api.internal;
 
+import java.util.Objects;
+import javax.annotation.Nullable;
 import javax.annotation.concurrent.Immutable;
 
-/** Internal utility methods for working with attribute keys, attribute values, and metric names. */
+/** Utilities for working with strings. */
```
[the remark this answers is not in the document]
Comment (kubawach) [reply api/all/src/main/java/io/opentelemetry/api/internal/StringUtils.java]: Good idea, done.

**Comment by jkwatson** [comment]
Looks like this is being bitten by whatever is causing the build failures from #3086  as well.

**Comment by kubawach** [comment]
> Looks like this is being bitten by whatever is causing the build failures from #3086 as well.

Yep - and locally the same test runs quite OK.

**Comment by jkwatson** [comment]
> > Looks like this is being bitten by whatever is causing the build failures from #3086 as well.
> 
> Yep - and locally the same test runs quite OK.

It looks like the failure is only on ubuntu, and everything is 100% ok on macos.

**Comment by kubawach** [comment]
> > > Looks like this is being bitten by whatever is causing the build failures from #3086 as well.
> > 
> > 
> > Yep - and locally the same test runs quite OK.
> 
> It looks like the failure is only on ubuntu, and everything is 100% ok on macos.

willing to take a risk and merge? 🤔 😀

**Comment by jkwatson** [comment]
> > > > Looks like this is being bitten by whatever is causing the build failures from #3086 as well.
> > > 
> > > 
> > > Yep - and locally the same test runs quite OK.
> > 
> > 
> > It looks like the failure is only on ubuntu, and everything is 100% ok on macos.
> 
> willing to take a risk and merge? 🤔 😀

I don't want to merge anything further to the main branch until we've got the CI issue fixed.

**Comment by kubawach** [comment]
> #3086

I see that it has been fixed, @jkwatson can you please kick rebuild on this PR?

**Comment by jkwatson** [comment]
> > #3086
> 
> I see that it has been fixed, @jkwatson can you please kick rebuild on this PR?

Where do you see it's been fixed?

**Comment by kubawach** [comment]
> > > #3086
> > 
> > 
> > I see that it has been fixed, @jkwatson can you please kick rebuild on this PR?
> 
> Where do you see it's been fixed?

I saw your comment about #3086 and somehow assumed this has been reverted. Yet I see it hasn't been yet (neither reverted nor fixed). So I'll wait then 🤷‍♂️

**Comment by jkwatson** [comment]
ok, now the breaking bit has been reverted. :)

File: api/all/src/main/java/io/opentelemetry/api/internal/StringUtils.java [hunk 4]
```diff
@@ -5,12 +5,78 @@
...
 
+import java.util.Objects;
+import javax.annotation.Nullable;
 import javax.annotation.concurrent.Immutable;
 
-/** Internal utility methods for working with attribute keys, attribute values, and metric names. */
+/**
+ * Utilities for working with strings.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change *
```
Comment (anuraaga) [inline api/all/src/main/java/io/opentelemetry/api/internal/StringUtils.java:15]:
```suggestion
 * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
```

**Pull request merged**